app/db/interface.py

```diff
@@ -139,12 +140,85 @@ def _remove_soft_deletes(self, payments):
-        but this will do for now 
+        but this will do for now
         """
         return [payment for payment in payments if not payment.is_removed]
-        
+
+    def perform_payment_action(self, user_id, payment_attributes=None):
+        """
+        Accepts a payment with user_id and performs an action on it
+
+        :param user_id: <int> the unique identifier of a user_id whose payment/s is/are to be actioned upon
+        :param payment_attributes: <dict> a collection of new payment attribute values that will update old ones
+
+        if payment_attributes has payment_id, then it is "set-default" action.
+        This also sets any other default payment of the same user to False
+
+        if payment_attributes has amount, then it is performing "charge" action.
+        This would update the charge history
+        """
+        try:
+            payments = self._query_payments(payment_attributes={"user_id":user_id})
+            if 'payment_id' in payment_attributes: #for set-default action
+                payment_id_to_be_updated = payment_attributes['payment_id']
+                payment_to_be_updated = None
+                for payment in payments:
+                    if payment.id == payment_id_to_be_updated: #no need to check if is_removed = False because the query filters it
+                        payment_to_be_updated = payment
+                if payment_to_be_updated != None:
+                    payment_to_be_updated.is_default = True
+                    #making sure other payments have default as false
+                    other_payments = [payment for payment in payments if payment != payment_to_be_updated]
+                    for other_payment in other_payments:
+                        other_payment.is_default = False
+                    #Committing all changes
+                    self.db.session.commit()
+                    return True
+                else:
+                    return False
+            elif 'amount' in payment_attributes: #for charge action
+                default_payment = None
+                for payment in payments:
+                    if payment.is_default == True:
+                        default_payment = payment
+                if default_payment != None:
+                    update_payment_flag = True
+
+                    if default_payment.payment_type == 'paypal' and not default_payment.details.is_linked:
+                        error_msg = 'Invalid request: Default Payment for this user_id: '+ str(user_id)+' (Paypal) is not linked.'
+                        raise DataValidationError(error_msg)
+                    elif default_payment.payment_type != 'paypal' and self.is_expired(default_payment.details.expires):
+                        error_msg = 'Invalid request: Default Payment for this user_id: '+ str(user_id)+' ('+default_payment.payment_type+') is expired'
+                        raise DataValidationError(error_msg)
```

Comment:
For the purposes of testing and consistency, it might be better to create constants for these two messages.  As for inserting the `user_id`, you can place a `{}` in the place where you want to insert the variable's value and then call `error_message.format(user_id)` - Python will figure out that the `user_id` value needs to go into where the `{}` is .